packages/lib/src/__tests__/logger.test.ts

```diff
@@ -0,0 +1,138 @@
+import { LoggingEvent } from 'log4js'
+
+import { Logger } from '@smartthings/core-sdk'
+
+import { logManager } from '../logger'
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const recording = require('log4js/lib/appenders/recording')
+
+
+describe('logger', () => {
+	function setupLogManager(level: string): void {
+		const config = {
+			appenders: { main: { type: 'recording' } },
+			categories: { default: { appenders: ['main'], level } },
+		}
+		logManager.init(config)
+	}
+
+	function logOneOfEach(log: Logger): LoggingEvent[] {
+		log.trace('trace message')
+		log.debug('debug message')
+		log.info('info message')
+		log.warn('warn message')
+		log.error('error message')
+		log.fatal('fatal message')
+
+		return recording.replay()
+	}
+
+	afterEach(() => {
+		recording.erase()
+		jest.clearAllMocks()
+	})
+
+	describe('logManager', () => {
+		// This test has to be first so logging hasn't been initialized for it.
+		it('getLogger throws exception if not initialized', function() {
+			expect(() => logManager.getLogger('my-category')).toThrow('logging not initialized')
+		})
+
+		it('returns the same logger for the same name', function() {
+			setupLogManager('error')
+			const one = logManager.getLogger('my-category')
+			const two = logManager.getLogger('my-category')
+			expect(one).toBe(two)
+		})
+
+		it('returns different loggers for different names', function() {
+			setupLogManager('error')
+			const one = logManager.getLogger('my-category')
+			const two = logManager.getLogger('your-category')
+			expect(one).not.toBe(two)
+		})
+
+		it('level returns correct level', function() {
+			setupLogManager('error')
+			const log = logManager.getLogger('my-category')
+
+			expect(log.level).toBe('ERROR')
+
+			setupLogManager('warn')
+			expect(log.level).toBe('WARN')
+		})
+
+		it('level sets the level correctly', function() {
+			setupLogManager('error')
+			const log = logManager.getLogger('my-category')
+
+			expect(log.level).toBe('ERROR')
+
+			log.level = 'warn'
+			expect(log.level).toBe('WARN')
+		})
+
+		it('logs correctly when level is trace', function() {
+			setupLogManager('trace')
+			const log = logManager.getLogger('my-category')
+			const logEvents = logOneOfEach(log)
+
+			expect(logEvents.length).toBe(6)
+		})
+
+		it('logs correctly when level is warn', function() {
+			setupLogManager('warn')
+			const log = logManager.getLogger('my-category')
+			const logEvents = logOneOfEach(log)
+
+			expect(logEvents.length).toBe(3)
+			expect(logEvents.map(event => event.level.levelStr)).toEqual(['WARN', 'ERROR', 'FATAL'])
+		})
+
+		it('log logs correctly', function() {
+			setupLogManager('trace')
+			const log = logManager.getLogger('my-category')
+
+			// We might consider removing this at some point...with log4js it
+			// seems to log at the info level.
```

Comment:
So are we saying that the test for `log` logging correctly in fact logs incorrectly?

Reply:
No, what I was really trying to say is that the `log` method that doesn't specify a level was itself a mistake. (What level _should_ it log at?) Logging frameworks commonly have a `log` method where you specify the level as an argument and when I wrote this in the first place, I proxied log4js methods and I believe I thought this was that method, not noticing the lack of a level specified. (Run-on enough for you? :-))

Reply:
I'll update the comment to make this more clear. Also, I would have removed this when I noticed it but it will require removing (or updating it) in the interface in the core SDK which deserves its own issue IMO.

Reply:
You know what, I"m just going to remove the comment entirely and write an issue instead. That's what I should have done in the first place. (Sorry for the reply-barrage. 😄 )